### PR TITLE
Yara 3.11.0

### PR DESCRIPTION
**Pull request by SaretX**
Update parser to YARA v.3.11.0

### CODE REVIEW

**Comment by Northern-Lights** [comment]
Thanks for doing this so quickly. I'm checking it out and will get back to you soon. I'll say upfront that I'm a little hesitant to use the `Number` type since that will be a breaking change, but it may be a decent solution to the number formatting issue.

**Comment by Northern-Lights** [comment]
The test rule `XOR_RANGE` does not seem to be proper syntax:
```
2020/03/25 20:04:22 Unable to parse ruleset file "ruleset.yar": Parser result: "1" grammar: lexical error @97: "syntax error: unexpected _XOR_, expecting _CONDITION_"
FAIL    github.com/Northern-Lights/yara-parser/tests    0.002s
FAIL
```
The `xor` modifier seems to be failing if it comes after a regex. I'm looking into this one...

As a separate issue, the ruleset also fails with yara on the command line:

```
> yara ruleset.yar ruleset.yar
ruleset.yar(1): error: can't open include file: ./true.yar
ruleset.yar(96): error in rule "XOR": invalid modifier combination "xor nocase"
ruleset.yar(107): error in rule "XOR": invalid modifier combination "xor nocase"
ruleset.yar(121): error in rule "XOR": syntax error, unexpected <xor>, expecting <condition>
```

Rules that we don't expect to compile should go into their own file so that results of the clean ruleset can continue to be evaluated

**Comment by Northern-Lights** [comment]
I also see that they've split up `string_modifiers` into `string_modifiers` and `regexp_modifiers` and `hex_modifiers`. Not sure if we should make some kind of `String` interface and have those 3 types implement it. Seems like it would work cleanly but might be kind of heavy...

**Comment by SaretX** [comment]
1:
> The test rule XOR_RANGE does not seem to pass
> The xor modifier seems to be failing if it comes after a regex.

This is an expected behavior, regexp strings are not supposed to work with XOR modifier, forgot to update test cases, my bad, fixed it in the update above.

2: 
>As a separate issue, the ruleset also fails with yara on the command line

Documentation is a bit vague on modifier combinations. In **nocase** section it claims:
> This modifier can be used in conjunction with any other modifier

but later in **xor** section:
>You can also combine the xor modifier with wide and ascii modifiers

I've added grammar check for combination of **xor** and **nocase** modifiers as it looks to be the only case when yara compiler throws an error.

3:
>I also see that they've split up string_modifiers into string_modifiers and regexp_modifiers and hex_modifiers. Not sure if we should make some kind of String interface and have those 3 types implement it

We already check provided modifiers for different types of strings in **grammar.y**, so it will throw an error in cases like combination of hex strings with any kind of modifier (except **private**) or **xor** modifier with regexp string. So I think that String interface doesn't make much sense and instead it's another breaking change and IMO much worse than Number type, as String type is probably much more used.

File: data/data.go [hunk 1]
```diff
@@ -58,11 +58,35 @@ const (
...
+
+// Type of number
+const (
+	Hex NumBase = iota
+	Decimal
+	Octal
+)
+
+// Number specifies range bound for strings XOR modifier
+type Number struct {
```
Comment (Northern-Lights) [inline data/data.go]: I can't get behind the `Number` type right now, as it's going to break some of my code, as well. It's an issue that I've been wanting to address, and I think that it deserves its own PR. Can you save this code and open a new PR for it?

**Comment by Northern-Lights** [comment]
Number-formatting proposal in #20

**Comment by Northern-Lights** [comment]
I created a 3.11.0 branch and merged it with my formatted numbers change. I'm still playing around with some XorRange alternatives to see if we can simplify the API.

File: data/data.go [hunk 2]
```diff
@@ -58,11 +58,35 @@ const (
...
-	Fullword bool `json:"fullword"`
-	Xor      bool `json:"xor"`
-	I        bool `json:"i"` // for regex
-	S        bool `json:"s"` // for regex
+	Nocase   bool     `json:"nocase"`
+	ASCII    bool     `json:"ascii"`
+	Wide     bool     `json:"wide"`
+	Fullword bool     `json:"fullword"`
+	Private  bool     `json:"private"`
+	Xor      bool     `json:"xor"`
```
[the remark this answers is not in the document]
Comment (SaretX) [reply data/data.go:66]: Yup, this is definitely a better way to implement XOR modifier.

**Comment by SaretX** [comment]
Sorry for the delayed response, had a few tough weeks here. I do like most of your changes, the only thing I'm concerned about is usage of **%w** verb that requires Go v1.13 to properly work and at the same time, most of those errors where you use it will never be thrown due to the appropriate checks in parser itself.

**Comment by Northern-Lights** [comment]
> Sorry for the delayed response, had a few tough weeks here. I do like most of your changes, the only thing I'm concerned about is usage of **%w** verb that requires Go v1.13 to properly work and at the same time, most of those errors where you use it will never be thrown due to the appropriate checks in parser itself.

Good point. I will revert those to %s before I tag another release.

**Pull request merged**